numpy/core/_asarray.py

```diff
@@ -293,14 +362,22 @@ def require(a, dtype=None, requirements=None):
       UPDATEIFCOPY : False
 
     """
+    if like is not None:
+        return _require_with_like(
+            a,
+            dtype=dtype,
+            requirements=requirements,
+            like=like,
+        )
+
     possible_flags = {'C': 'C', 'C_CONTIGUOUS': 'C', 'CONTIGUOUS': 'C',
                       'F': 'F', 'F_CONTIGUOUS': 'F', 'FORTRAN': 'F',
                       'A': 'A', 'ALIGNED': 'A',
                       'W': 'W', 'WRITEABLE': 'W',
                       'O': 'O', 'OWNDATA': 'O',
                       'E': 'E', 'ENSUREARRAY': 'E'}
     if not requirements:
-        return asanyarray(a, dtype=dtype)
```

Comment:
Why pass like here given we know it is none? Do we actually want to dispatch to `like` in these functions, or should we just forward the argument to `np.array`?

Reply:
Good catch, this was probably a leftover from some early attempts, I now removed those in https://github.com/numpy/numpy/pull/16935/commits/c1e87547a2edaa08a2bc909a6ce6367b277ebb26 .

Reply:
I think this pattern is present in lots of places. I'd perhaps argue that for instance `np.array` should have `like` dispatching, but `np.asarray` and `np.asanyarray` should remain an alias for `array(subok=..., copy=...)`. Thoughts @seberg?

Reply:
Downstream libraries may have completely different implementations, that's why we need each function to have `like=`. For instance, CuPy doesn't support `subok`, as per the note in https://docs.cupy.dev/en/stable/reference/generated/cupy.array.html , so we have to dispatch it directly to CuPy so it handles that appropriately.

Reply:
Hmmmpf, I thought I had brought that up but forgot about it again.

I am not sure why we should have more than one entry point for the `as.*array()` functions.  In principle it seems like only ever using `np.array()` with the full signature during `__array_function__` dispatching should be good enough?

 @pentschev did you have a reason for wanting `np.asarray`, `np.ascontiguousarray`, ... instead of only `np.array()` as the `public_api` dispatching "target"?  Dask should not really have to know that the call was `np.asarray(arr)` as opposed to `np.array(arr, copy=False)`, right?  And that would mean fewer functions to wrap (although I guess you probably just wrap everything anyway).

We might still want an `if like is None:` branch for micro-optimization, but I am not sure whether it will make a real difference in that case.

Reply:
@seberg did you see my comment just above https://github.com/numpy/numpy/pull/16935#discussion_r477640294 ? Maybe GH didn't refresh it while you were typing, as it happened several times with me. :)

Reply:
Yeah, I was typing when you posted.  That seems like a reason, although I didn't quite grok the implications yet...

Reply:
Note we also discussed this previously in https://github.com/numpy/numpy/pull/16935#issuecomment-670123319 : 

> > Or, we admit that these are just aliases around `np.array`, and only defer for `np.array` itself. But I suppose we do not want that, since the error messages would be confusing (`np.asarray` saying that it did not find an implementation for `np.array`).
> 
> I did that in the beginning, but I realized that's not necessarily correct and removed that in [2c54820](https://github.com/numpy/numpy/commit/2c5482058929e592b74edf5228761780d60e931a), downstream libraries may implement these functions in a completely different manner than NumPy.

Reply:
> Yeah, I was typing when you posted. That seems like a reason, although I didn't quite grok the implications yet...

The implication is when you call `np.array(..., subok=True, like=cupy_array)` would will fail with `NotImplemented` because `cupy.array` doesn't implement `subok`, and `cupy.asanyarray` is implemented differently which works with how it's currently implemented doing `np.asanyarray(..., like=cupy_array)`.

Reply:
Hmm, yeah, I don't mind these aliases too much (they are a bit tedious, but only a bit), but I don't really *like* it much either :). While `cupy.array()` probably shouldn't support `subok=True`, I wonder if the function called by its `__array_function__` could just as well support (and ignore) it...

Although the whole subok parameter is slightly dubious in combination with `like=`?

PS: I added it to the tracking issue just in case we don't quite find a resolution.

Reply:
As far as I understand it, implementing the NumPy API doesn't imply that the underlying implementation has to behave exactly the same. With that said, some downstream library implementing `asanyarray` could, for example, implement it to always return `np.array(1)` for whatever reason, and that could be totally valid. Therefore, from the NumPy perspective, we shouldn't assume that downstream libraries are supposed to implement those aliases exactly the same way as NumPy does. My example is obviously a bit extreme, but I'm just trying to make the point that with `__array_function__` we're using the API only, and the underlying implementation is entirely of responsibility of the downstream library and thus we should only dispatch on the API.